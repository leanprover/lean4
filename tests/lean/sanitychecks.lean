theorem unsound : False := -- Error
  unsound

partial theorem unsound2 : False := -- Error
  unsound2

unsafe theorem unsound3 : False := -- Error
  unsound3

opaque unsound4 : False  -- Error

axiom magic : False -- OK
namespace Foo
partial def foo (x : Nat) : Nat := foo x  -- OK

unsafe def unsound2 : False := unsound  -- OK

partial def unsound3 : False := unsound3  -- Error

partial def unsound4 (x : Unit) : False := unsound4 ()  -- Error

partial def badcast1 (x : Nat) : Bool :=
  unsafeCast x -- Error: partial cannot use unsafe constant

partial def badcast2 (x : Nat) : Bool :=
  if x == 0 then unsafeCast x -- Error: partial cannot use unsafe constant
  else badcast2 (x + 1)

unsafe def badcast3 (x : Nat) : Bool := -- OK
  if x == 0 then unsafeCast x
  else badcast3 (x + 1)
