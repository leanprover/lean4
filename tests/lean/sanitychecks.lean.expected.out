sanitychecks.lean:1:8-1:15: error: fail to show termination for
  unsound
with errors
structural recursion cannot be used

well-founded recursion has not been implemented yet
sanitychecks.lean:4:8-5:9: error: 'partial' theorems are not allowed, 'partial' is a code generation directive
sanitychecks.lean:7:7-8:9: error: 'unsafe' theorems are not allowed
sanitychecks.lean:10:0-10:24: error: failed to synthesize instance
  Inhabited False
sanitychecks.lean:18:0-18:40: error: failed to compile partial definition 'unsound3', failed to show that type is inhabited
sanitychecks.lean:20:12-20:20: error: (kernel) invalid declaration, it uses unsafe declaration 'unsafeCast'
sanitychecks.lean:23:12-23:20: error: (kernel) invalid declaration, it uses unsafe declaration 'unsafeCast'
